hi-ml/testhiml/testhiml/test_run_ml.py

```diff
@@ -220,11 +220,41 @@ def test_run_training() -> None:
             mock_trainer.loggers[0].finalize.assert_called_once()
 
 
+@pytest.mark.parametrize("max_num_gpus_inf", [-1, 1])
+def test_end_training(max_num_gpus_inf: int) -> None:
+    experiment_config = ExperimentConfig(model="HelloWorld")
+    container = HelloWorld()
+    container.max_num_gpus_inference = max_num_gpus_inf
+    runner = MLRunner(experiment_config=experiment_config, container=container)
+
+    with patch.object(container, "get_data_module"):
+        with patch("health_ml.run_ml.create_lightning_trainer", return_value=(MagicMock(), MagicMock())):
+            runner.setup()
+            runner.init_training()
+            runner.run_training()
+
+        with patch.object(runner.checkpoint_handler, "additional_training_done") as mock_additional_training_done:
+            with patch.object(runner, "after_ddp_cleanup") as mock_after_ddp_cleanup:
+                with patch("health_ml.run_ml.cleanup_checkpoints") as mock_cleanup_ckpt:
+                    old_environ = {"old": "environ"}
+                    runner.end_training(old_environ=old_environ)
+                    mock_additional_training_done.assert_called_once()
+                    mock_cleanup_ckpt.assert_called_once()
+                    if max_num_gpus_inf == 1:
```

Comment:
Maybe we could also call "assert_not_called" if max_num_gpus_inf is not equal to 1? Also, above we have another value, -1, for max_num_gpus_inf - what is the expected behaviour there? We don't seem to test that case here

Reply:
that's a very good point, I think assert_not_called would capture the expected behaviour nicely, that means if the max_num_gpus_inf != 1 we don't want to kill ddp processes and carry on with validation and test